README.md
# Jellyfish
Monitors the health and status of applications running in Marathon or individual services.

## Mode of operation
Jellyfish will visit all configured marathons asynchronously, one thread for every marathon,
  get all the apps and then visit their status page.

All individual configured services will be processed on one extra thread.

If aws credentials are configured, jellyfish will ask AWS Beanstalk for all of its environments and will monitore them.
Because Beanstalk does not necessarily follow the same naming conventions as marathon, you have to specify to which namespace the Beanstalk services belong (see configuration example).

## Status Level
Jellyfish knows four status levels:

- 0) **OK**: Everything is fine.
- 1) **UNKNOWN**: App is suspended or Status Page is not available, but app is 'healthy' for Marathon.
- 2) **WARNING**: App is in status 'WARNING'.
- 3) **ERROR**:	App is in status 'ERROR' or there are fewer healthy instances than there should be.

## Job Status Age
If you supply a 'stopped' field in your job information Jellyfish will display the status age (for status other than OK).

## URL Parameter:
### level=[0-3]
Filters applications and jobs by status and only displays application and jobs who met the definded status level or above (worse).

Example: http://jellyfish.com/monitor?level=2

#### jobs=[true/false]
If true, jobs will be displayed. Default is true.

Example: http://jellyfish.com/monitor?jobs=false

#### active_color_only=[true/false]
Show only active color. Default is false.

Example: http://jellyfish.com/monitor?active_color_only=true

#### Filter
The following filter all support comma separated lists. If you want to exclude something, just add an leading **!**.   

Positive filtering always wins, because it excludes everything else.

Example: http://jellyfish.com/monitor?filter=service1,service2

##### filter=[app_name, !app_name]
Filters application names.

##### group=[group_name, !group_name]
Filters groups.

##### type=[type, !type]
Filter for the value of the marathon label "type".

##### env=[environment, !environment]
Filters environments aliase.

#### refresh=[seconds]
Adds metatag to trigger a page refresh every <seconds>.

Example: http://jellyfish.com/monitor?refresh=5

## Dashboard Mode
There is a dashboard mode: /monitor/cinema.

## Severity rating and sorting
Services are sorted by their severity rating, which is calculated from all instances of one service.

Live problems are always more severe than non-live problems (Status Rating multiplied by 100).
 
Errors are always more severe than any other status (Status Rating multiplied by 10).

## Configuration

Example:

````
    environments:
      - name: develop
        alias: DEV
      - name: live
        alias: LIVE
    defaults:
      marathons:
        protocol: http
        apps: /v2/apps
        username: some_user
        password: some_password
        root_app_lable: ROOT_LABEL
        status_path_lable: STATUS_PATH_LABEL
        base_domain: some-domain.de
          - ".*marathon-healthcheck"
        graphite:
          - mem: "http://graphite.some-query.de"
          - cpu: "http://graphite.some-query.de"
    marathons:
      - host: marathon.pete.com
      - host: marathon.josh.com
        cookies:
          CookieName: CookieValue
        status_path:
          service_name: https://status_path
    services:
      - id: /develop/vertical/service
        url: http://some_url/status
    aws:
    - id: /develop/pete
      access_key: AAAA
      secret_key: 1234
````
The *defaults* will be merged with the following configuration.

In this example, everything under *defaults/marathons* will be merged with every map below *marathons*.

For more information visit the [Eliza dokumentation](http://eliza.readthedocs.io/en/latest/)

### Vault or Env
To read from Vault or Env use one of the following:

<%= VAULT['secret_path'] %>

<%= ENV['env_name'] %>

## Getting Started

### Obtaining it from source:

````bash
    $ git clone git@github.com:otto-de/jellyfish.git
````

### Installing from source:

````bash
    $ ./setup.sh [OPTIONS]
````
    Options:
        --no-venv   Installs packages globally. May need root privileges.  
        --doc-deps  Includes depedencies to generate documentation.

### Run:

````bash
    $ ./run.py [-h] [-p [PORT]] [-e [ENV]] [-w [WORKDIR]] [-g] [-v [VERBOSE]]

    optional arguments:
      -h, --help                
                            show help message and exit
      -p [PORT], --port [PORT]  
                            Port: Specify port. Default is 8080
      -e [ENV], --env [ENV]     
                            Environment: Specify which config to load. Default is local.
      -w [WORKDIR], --workdir [WORKDIR]
                            Workdir: Specify which working directory to use.
                            Default is the local directory
      -g, --greedy          
                            Greedy: Run processes once (synchron) and then start
                            to serve.
      -v [VERBOSE], --verbose [VERBOSE]
                            Lets you set the loglevel. Application default: ERROR.
                            Option default: INFO
````

### Run tests:

````bash
    $ ./run-tests.sh <environment>
````

## Contribute

Jellyfish is currently in active development and welcomes code improvements, bug fixes, suggestions and feature
requests. 

For those of your interested, providing documentation to other parties is equally welcome.

Please document all notable changes to this project in the provided changelog. Note that this project adheres to [Semantic Versioning](http://semver.org/).

## License

Distributed under the Apache License 2.0


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](https://github.com/otto-de/jellyfish/compare/1.1.0...HEAD)

## [1.1.0](https://github.com/otto-de/jellyfish/compare/1.0.1...1.1.0) - 2018-02-16
### Add
- Module to include AWS Beanstalk services.

### Fix
- Services with same name from different modules should not overwrite each other.

## [1.0.1](https://github.com/otto-de/jellyfish/compare/1.0.0...1.0.1) - 2017-05-30
### Fix
- Service order is stable (lexical order), even if severity ratings are equal.

## [1.0.0](https://github.com/otto-de/jellyfish/compare/0.15.1...1.0.0) - 2017-05-19
### Removed
- resource view feature was removed.

## [0.15.1](https://github.com/otto-de/jellyfish/compare/0.15.0...0.15.1) - 2017-03-27
### Fixed
- Properly convert root_app flag into boolean. 


## [0.15.0](https://github.com/otto-de/jellyfish/compare/0.14.0...0.15.0) - 2017-03-2
### Add
- Severity rating and sorting of services.

## [0.14.0](https://github.com/otto-de/jellyfish/compare/d0ff089a5409d9e7e00150b381e9c01a34bb9e5d...0.14.0) - 2016-11-9
### Add
- Tidy up resource overview.
- Mini pie charts for resource overview.

## 0.13.1 - 2016-11-7
### Fixed
- Some jobs do not have messages.

## 0.13.0 - 2016-11-7
### Changed
- Use lable names for status path and root app from config.
- Use domain name from config.

## 0.12.0 - 2016-10-31
### Changed
- Use full name for all tab to distinguish between services with the same name from multiple verticals.

## 0.11.1 - 2016-10-4
### Fixed
- Add excluding filter support for environment filter 

## 0.11.0 - 2016-09-28
### Added
- Parameter to include auto refresh metatag.
- App name filter now supports excluding services with leading *!* (example: filter=!service).
- Filter for verticals / groups (example: group=p13n,reco)
- Filter for marathon type label.

## 0.10.0 - 2016-09-26
### Added
- Cookie support for status page query.
- Configurable header for status page query.
- Configurable blacklist for services.

## 0.9.0 - 2016-09-22
Some general refactoring.

### Added
- Support for single services outside of marathon.
- Running jobs will be represented by blinking/glowing status icons.
- Styleguide html test.
- Resource Allocation tab. 

## 0.8.0 - 2016-08-15
### Added [experimental]
- Displays peak resource usage of the last 14 days.

## 0.7.1 - 2016-08-12
### Fixed
- Use all found apps to calculate resource allocation, instead of only the filtered ones.

## 0.7.0 - 2016-08-12
### Changed
- Jellyfish does not save previous job status to display the status age anymore, instead it will use the 'stopped timestamp' from the status page (as of Edison-Microservice version 0.69.1).

## 0.6.0
### Added
- Display resource allocation for each vertical, task and application.
- A service will be marked as "ERROR" if the status code of the status page request is >= 500. This is to indicate infrastructure problems, where the application is healthy, but not accessible.

### Changed
- Use Root_App flag to build status path to build the correct status path for services with names that differ from the vertical name.
- Environment Columns should always be the same size.
- Increase size of status indicator (10px -> 15px).

## 0.5.1
### Fixed
- Status age button sets correct parameter.

## 0.5.0
### Added
- Display job messages as tooltip.
- TEST FEATURE: Display job status age (from jellyfish's point of view) for jobs != OK.

### Changed
- Logging will include error types for better analysis.

## 0.4.3 - 2016-07-13
### Fixed
- Fix Query Strings forwarding.

## 0.4.1 - 2016-07-12
### Fixed
- Parameter handling.

## 0.4.0 - 2016-07-12
### Added
- 'all' tab to include all services.
- 'vertical' tag for every service on 'all' tab to identify their source.
- Service counter on tabs.

## 0.3.0 - 2016-07-12
### Added
- Support Marathon subgroups.
- Support status path information in lables.
- Styleguide link.
- Use python argument parser for startup.

### Changed
- Use marathon hostname to identify thread.
- Change button semantics (current state instead of action, eg. Hide jobs -> Jobs hidden).
- Change some ui wording.

### Fixed
- Show empty tab is filter result is empty.
- Handle Marathon not responding.
- Styleguide.

## 0.2.0 - 2016-07-08
### Added
- Show Active only colors.

### Changed
- Buttons now show the actual state instead of action ("Hide Jobs" -> "Jobs hidden"). The "closed eye" icon symbolised that information is hidden from you.

### Issues:
- Status page of all colors of an app identical.

## 0.1.0 - 2016-07-08
First release of Jellyfish.

### Added
- Display status of all apps in Marathon.
- Status level filter.
- Job filter (true/false).
- Name filter.
- Size (heading) parameter.
- Link to Marathon.
- Link to status page.

### Issues:
- Status page of all colors of an app identical.

This changelog is inspired by [keepachangelog.com](http://http://keepachangelog.com/de/)


app/update.py
# -*- coding: utf-8 -*-
import json
import pickle
import re
from pprint import pprint
from threading import Timer
from urllib.parse import quote_plus

import requests
from delorean import Delorean
from flask import logging

from app import config

logger = logging.getLogger(__name__)


def update_marathon(thread_id, cfg, interval, greedy=False):
    app_list = get_apps(cfg)
    previous_tasks = (config.rdb.get(thread_id) or b'').decode().split(',')
    blacklist = '(?:%s)' % '|'.join(cfg['blacklist']) if 'blacklist' in cfg else '1234'
    tasks = list()
    for app in app_list:
        try:
            _, _, _, app_name, _ = itemize_app_id(app["id"])
            if not re.match(blacklist, app["id"]):
                config.rdb.sadd("all-services", app["id"])
                config.rdb.set(app["id"], json.dumps(get_task_info(app, cfg)))
                tasks.append(app["id"])
        except Exception as error:
            logger.error(
                ' '.join([cfg['host'] + ":" + (app["id"] if app else "unkown"), "[",
                          error.__class__.__name__, "]"]), exc_info=True)

    for app_id in list(set(previous_tasks) - set(tasks)):
        config.rdb.delete(app_id)
        config.rdb.srem("all-services", app_id)
    config.rdb.set(thread_id, ",".join(tasks))

    config.rdb.set(cfg['host'], pickle.dumps(Delorean.now()))
    if not greedy:
        logger.debug("Finish update for " + cfg['host'])
        Timer(interval=interval, function=update_marathon,
              args=(thread_id, cfg, interval)).start()


def update_service(thread_id, service_list, interval, greedy=False):
    for service in service_list:
        config.rdb.sadd("all-services", service["id"])
        config.rdb.set(service["id"], json.dumps(get_service_info(service)))
    config.rdb.set("single_services", pickle.dumps(Delorean.now()))
    if not greedy:
        logger.debug("Finish update for services")
        Timer(interval=interval, function=update_service,
              args=(thread_id, service_list, interval)).start()


def update_aws(thread_id, service, interval, greedy=False):
    application_environment_mapping = get_beanstalk_environments(service)
    for application, environment in application_environment_mapping.items():
        health = get_beanstalk_health(environment)
        app_id = service["id"] + '/' + application
        config.rdb.set(app_id, json.dumps(health))
        config.rdb.sadd("all-services", app_id)
    if not greedy:
        logger.debug("Finish update for aws")
        Timer(interval=interval, function=update_service,
              args=(thread_id, service, interval)).start()


def get_apps(marathon):
    marathon_url = ''.join([marathon['protocol'],
                            "://",
                            quote_plus(marathon['username']),
                            ':',
                            quote_plus(marathon['password']),
                            '@',
                            marathon['host'],
                            marathon['apps']])
    try:
        request_result = requests.get(marathon_url,
                                      headers={"Accept": "application/json"},
                                      verify=False,
                                      timeout=5)
        data = json.loads(request_result.text)
        config.rdb.set(marathon['host'] + '-cache', json.dumps(data["apps"]))
        config.rdb.delete(marathon['host'] + '-errors')
    except (ValueError, requests.exceptions.Timeout,
            requests.exceptions.ConnectionError) as error:

        error_message = ' '.join(["could not read marathon:", marathon['host'], "[", error.__class__.__name__, "]"])

        logger.warning(error_message)
        config.rdb.set(marathon['host'] + '-errors', error_message)
        raw_cache = config.rdb.get(marathon['host'] + '-cache')
        if raw_cache:
            return json.loads(raw_cache.decode())
        return []
    return data["apps"]


def get_task_info(app, cfg):
    task = dict()
    task["id"] = app["id"]
    task["group"], task["vertical"], task["subgroup"], task["name"], task["color"] = itemize_app_id(app["id"])
    task["full-name"] = "-".join([task["vertical"], task["name"]])

    task["marathon"] = dict()
    task["marathon"]["origin"] = cfg["host"]
    task["marathon"]["instances"] = app["instances"]
    task["marathon"]["cpu"] = app["cpus"]
    task["marathon"]["mem"] = app["mem"]
    task["marathon"]["staged"] = app["tasksStaged"]
    task["marathon"]["running"] = app["tasksRunning"]
    task["marathon"]["healthy"] = app["tasksHealthy"]
    task["marathon"]["unhealthy"] = app["tasksUnhealthy"]
    task["marathon"]["marathon_link"] = cfg['protocol'] + "://" + cfg['host'] + "/ui/#/apps/" + \
                                        quote_plus(task["id"])
    task["marathon"]["labels"] = app["labels"]

    root_app = get_in_dict(["env", cfg.get("root_app_lable")], app, "") or \
               get_in_dict(["labels", cfg.get("root_app_lable")], app, "")
    root_app = root_app == 'true' or root_app == 'True'
    status_path = get_in_dict(["env", cfg.get("status_path_lable")], app, "") or \
                  get_in_dict(["labels", cfg.get("status_path_lable")], app, "")
    task["status_url"] = get_status_url(task["name"], task["group"], task["vertical"], task["subgroup"],
                                        cfg['base_domain'], status_path,
                                        root_app, cfg)

    if status_path and task["marathon"]["instances"] > 0:
        status_page_data, active_color, status_page_code = get_application_status(task["status_url"], cfg)
        if task["color"] and active_color and task["color"] != active_color:
            task["status_url"] = task["status_url"].replace('http://', 'http://staged.')
            status_page_data, _, status_page_code = get_application_status(task["status_url"], cfg)
        task["version"] = get_in_dict(["application", "version"], status_page_data, "UNKNOWN")
        task["status_page_status_code"] = status_page_code
        task["active_color"] = active_color
        task["app_status"] = status_level(get_in_dict(["application", "status"], status_page_data, "UNKNOWN"))
        task["jobs"] = dict()
        for job, job_info in get_in_dict(["application", "statusDetails"], status_page_data, {}).items():
            task["jobs"][job] = get_job_info(job_info)
    else:
        task["version"] = "UNKNOWN"
        task["active_color"] = None
        task["status_page_status_code"] = None
        task["app_status"] = status_level("UNKNOWN")
        task["jobs"] = dict()

    task["status"] = overall_status(task)
    task["severity"] = calculate_severity(task)
    return task


def get_service_info(service):
    task = dict()
    task["id"] = service["id"]
    task["status_url"] = service["url"]
    task["group"], task["vertical"], task["subgroup"], task["name"], task["color"] = itemize_app_id(service["id"])
    task["full-name"] = "-".join([task["vertical"], task["name"]])

    status_page_data, active_color, status_page_code = get_application_status(service["url"], service)
    task["version"] = get_in_dict(["application", "version"], status_page_data, "UNKNOWN")
    task["status_page_status_code"] = status_page_code
    task["active_color"] = active_color
    task["app_status"] = status_level(get_in_dict(["application", "status"], status_page_data, "UNKNOWN"))
    task["jobs"] = dict()
    for job, job_info in get_in_dict(["application", "statusDetails"], status_page_data, {}).items():
        task["jobs"][job] = get_job_info(job_info)

    task["status"] = task["app_status"]
    task["severity"] = calculate_severity(task)
    return task


def get_beanstalk_environments(service):
    return {}


def get_beanstalk_health(service):
    return {}


def get_status_url(name, group, vertical, subgroup, base_domain, status_path, root_app, cfg):
    if 'status_path' in cfg and name in cfg['status_path']:
        return cfg['status_path'][name].replace('{environment}', group)
    if root_app:
        name = vertical
    if status_path:
        status_url_prefix = name + "." + subgroup if subgroup else name
        status_url_prefix = status_url_prefix + "." + vertical if not root_app else name
        return ''.join(['http://', status_url_prefix, '.', group, '.', base_domain, status_path])
    else:
        return ''


def get_application_status(status_url, cfg):
    status_code = None
    try:
        request_result = requests.get(status_url,
                                      headers=cfg['headers'] if 'headers' in cfg else {},
                                      cookies=cfg['cookies'] if 'cookies' in cfg else {},
                                      verify=False,
                                      timeout=5)
        active_color = request_result.headers.get('x-color', None)
        status_code = request_result.status_code
        status_page_data = json.loads(request_result.text)
    except (ValueError, requests.exceptions.Timeout,
            requests.exceptions.ConnectionError) as error:
        logger.warning(' '.join(["could not read status page:", status_url, "[", error.__class__.__name__, "]"]))
        return {}, None, status_code
    return status_page_data, active_color, status_code


def get_in_list(index, list, default=None):
    if len(list) > index:
        return list[index]
    else:
        return default


def get_in_dict(key_list, my_dict, default=None):
    tmp = my_dict
    for key in key_list:
        if not isinstance(tmp, dict):
            return default
        tmp = tmp.get(key, default)
    return tmp


def itemize_app_id(app_id):
    color_codes = ['BLU', 'GRN']
    split = app_id.split("/")
    group = split[1]
    vertical = split[2]

    if len(split) == 4 or len(split) == 5 and str.upper(split[-1]) in color_codes:
        subgroup = ""
        name = split[3]
    else:
        subgroup = split[3]
        name = split[4]

    if str.upper(split[-1]) in color_codes:
        color = str.upper(split[-1])
    else:
        color = 'GRN'
    return group, vertical, subgroup, name, color


def calculate_severity(task):
    return task["status"] * (10 if task["status"] == 3 else 1) * (100 if task['group'] == 'live' else 1)


def overall_status(task):
    if task["marathon"]["healthy"] < task["marathon"]["instances"] or \
                    task["status_page_status_code"] and task["status_page_status_code"] > 500:
        return status_level("ERROR")
    else:
        return 1 if "app_status" not in task or task["app_status"] is None else task["app_status"]


def get_job_info(job):
    current_status = status_level(job["status"])
    return {"status": current_status,
            "message": job.get("message", ""),
            "running": bool("running" in job and job["running"]),
            "started": job.get("started", None),
            "stopped": job.get("stopped", None)}


def status_level(status):
    level = 1
    if status == "OK":
        level = 0
    if status == "UNKNOWN":
        level = 1
    if status == "WARNING":
        level = 2
    if status == "ERROR":
        level = 3
    return level
